pkg/term/winconsole/console_windows.go

```diff
@@ -0,0 +1,1124 @@
+func GetHandleInfo(in interface{}) (uintptr, bool) {
+	var inFd uintptr
+	var isTerminalIn bool
+	if file, ok := in.(*os.File); ok {
+		inFd = file.Fd()
+		isTerminalIn = IsTerminal(inFd)
+	}
+	if tr, ok := in.(*terminalReader); ok {
+		if file, ok := tr.wrappedReader.(*os.File); ok {
+			inFd = file.Fd()
+			isTerminalIn = IsTerminal(inFd)
+		}
+	}
+	return inFd, isTerminalIn
+}
+
+// GetConsoleMode gets the console mode for given file descriptor
+// http://msdn.microsoft.com/en-us/library/windows/desktop/ms683167(v=vs.85).aspx
+func GetConsoleMode(fileDesc uintptr) (uint32, error) {
+	var mode uint32
+	err := syscall.GetConsoleMode(syscall.Handle(fileDesc), &mode)
+	return mode, err
+}
+
+// SetConsoleMode sets the console mode for given file descriptor
+// http://msdn.microsoft.com/en-us/library/windows/desktop/ms686033(v=vs.85).aspx
+func SetConsoleMode(fileDesc uintptr, mode uint32) error {
+	r, _, err := setConsoleModeProc.Call(fileDesc, uintptr(mode), 0)
+	if r == 0 {
+		if err != nil {
+			return err
+		}
+		return syscall.EINVAL
+	}
+	return nil
+}
+
+// SetCursorVisible sets the cursor visbility
+// http://msdn.microsoft.com/en-us/library/windows/desktop/ms686019(v=vs.85).aspx
+func SetCursorVisible(fileDesc uintptr, isVisible BOOL) (bool, error) {
+	var cursorInfo CONSOLE_CURSOR_INFO
+	r, _, err := getConsoleCursorInfoProc.Call(uintptr(fileDesc), uintptr(unsafe.Pointer(&cursorInfo)), 0)
+	if r == 0 {
+		if err != nil {
+			return false, err
+		}
+		return false, syscall.EINVAL
+	}
+	cursorInfo.Visible = isVisible
+	r, _, err = setConsoleCursorInfoProc.Call(uintptr(fileDesc), uintptr(unsafe.Pointer(&cursorInfo)), 0)
+	if r == 0 {
+		if err != nil {
+			return false, err
+		}
+		return false, syscall.EINVAL
+	}
+	return true, nil
+}
+
+// SetWindowSize sets the size of the console window.
+func SetWindowSize(fileDesc uintptr, width, height, max SHORT) (bool, error) {
+	window := SMALL_RECT{Left: 0, Top: 0, Right: width - 1, Bottom: height - 1}
+	coord := COORD{X: width - 1, Y: max}
+	r, _, err := setConsoleWindowInfoProc.Call(uintptr(fileDesc), uintptr(BOOL(1)), uintptr(unsafe.Pointer(&window)))
+	if r == 0 {
+		if err != nil {
+			return false, err
+		}
+		return false, syscall.EINVAL
+	}
+	r, _, err = setConsoleScreenBufferSizeProc.Call(uintptr(fileDesc), uintptr(marshal(coord)))
+	if r == 0 {
+		if err != nil {
+			return false, err
+		}
+		return false, syscall.EINVAL
+	}
+
+	return true, nil
+}
+
+// GetConsoleScreenBufferInfo retrieves information about the specified console screen buffer.
+// http://msdn.microsoft.com/en-us/library/windows/desktop/ms683171(v=vs.85).aspx
+func GetConsoleScreenBufferInfo(fileDesc uintptr) (*CONSOLE_SCREEN_BUFFER_INFO, error) {
+	var info CONSOLE_SCREEN_BUFFER_INFO
+	r, _, err := getConsoleScreenBufferInfoProc.Call(uintptr(fileDesc), uintptr(unsafe.Pointer(&info)), 0)
+	if r == 0 {
+		if err != nil {
+			return nil, err
+		}
+		return nil, syscall.EINVAL
+	}
+	return &info, nil
```

Comment:
this feels wrong, if `r != 0`, according to MSDN docs, it means it's an error. However, we return no error.